LibertyDSNP/example-client
# Lazy load activity content

Purpose
---------------
Currently the example client attempts to load all existing content on the DSNP network at startup. Obviously, this is not going to scale. This PR reworks the app pretty significantly (but mostly behind the scenes), to load content only when we need to show it. Specifically it introduces and "Infinite scroll" feature to the feed to display older posts only after scrolling down to them. It then lazy loads all the activity content (posts, replies and profiles) so they are only loaded when viewed.


Solution
---------------
The first thing to note is that this PR introduces `react-query` to fetch activity content. This is somewhat controversial because this project already uses redux, so react-query is in some ways a second store. The motivation is that we want to load activity content only when it is viewed, but it may be viewed in multiple places simultaneously and will be re-rendered often. We could try to manage this by saving the content in redux, but we would need to worry about cache invalidation and we'd end up adding a lot of logic that does exactly what react-query does. It's better to think of it like the UI always goes out to the internet for activity content, only we put a good caching layer in front of it.

Another thing to note is that the example client had been doing several things wrong when it comes to announcements. This PR was forced to fix those.


Change summary
---------------
* Fix populate script to generate announcements of the correct type.
* Introduce a react-query QueryProvider at the top of the app.
* Modify FeedItem to contain only announcement information.
* Generally replace `Profile` (ProfileAnnouncement | ActivityContentProfile) with `User` (ProfileAnnouncement).
* Modify subscription and redux logic to only store announcements without first loading the activity content.
* Introduce separate `replies` structure in feed slice, so replies can be stored separately from posts and be more easily retrieved by inReplyTo field.
* Introduce infinite scroll component in PostList.
* Load activityContent of reply in Reply component.
* Load activityContent of post in the Post component.
* Modify `UserAvatar` to take a `User` and load the profile with the icon.
* Extract FromName component and a `userIdentifier` function to standardize logic of naming users that may not have a profile or handle.
* Fix announcement creation in sendPost, sendReply and saveProfile.
* Miscellaneous cleanup. 
* Fix tests.


Steps to Verify
----------------
1. Set up a clean populated app.
2. Note that everthing still works.
3. In the network tab, you might notice feed items loading as you scroll down the page. You probably won't though because the app tends to make all feed items visible briefly as they load from the subscription, which completely primes the react-query cache.
